pkg/controller/windowsmachine/metrics/metrics.go

```diff
@@ -0,0 +1,210 @@
+		windowsMetricsEndpoints, metav1.GetOptions{})
+	if err != nil {
+		return "", errors.Wrapf(err, "error getting endpoints object %v", windowsMetricsEndpoints)
+	}
+
+	foundNode := false
+	nodePosition := -1
+	for index, endpointAddress := range endpoints.Subsets[0].Addresses {
+		if pc.hostname == endpointAddress.TargetRef.Name {
+			foundNode = true
+			nodePosition = index
+			break
+		}
+	}
+
+	if !foundNode {
+		return "", errors.New(fmt.Sprintf("target node %s not found in Endpoints object %s",
+			pc.hostname, endpoints.Name))
+	}
+
+	return strconv.Itoa(nodePosition), nil
+}
+
+// addEndpointAddress adds node ip address to the list of subsets in Endpoints object
+func (pc *prometheusNodeConfig) addEndpointAddress() error {
+	_, err := pc.isNodeInEndpoints()
+	if err != nil {
+		// Add node if not present in the endpoints list
+		patchData := fmt.Sprintf("[{\"op\": \"add\", \"path\": \"/subsets/0/addresses/-\", \"value\":"+
+			" {\"ip\": \"%s\", \"targetRef\": {\"kind\": \"Node\", \"name\":\"%s\"}}}]",
+			pc.endpointAddress.IP, pc.hostname)
+
+		_, err := pc.k8sclientset.CoreV1().Endpoints(pc.namespace).Patch(context.TODO(),
+			windowsMetricsEndpoints, types.JSONPatchType, []byte(patchData), metav1.PatchOptions{})
+		if err != nil {
+			return errors.Wrapf(err, "could not add node %v to endpoints object, %v", pc.hostname, pc.endpointAddress.IP)
+		}
+		log.Info("prometheus configured", "operation", "add", "IPAddress", pc.endpointAddress.IP)
+		return nil
+	}
+	return nil
+}
+
+// removeEndpointAddress removes node ip address to the list of subsets in Endpoints object
+func (pc *prometheusNodeConfig) removeEndpointAddress() error {
+	nodeIPPosition, err := pc.isNodeInEndpoints()
+	// If node not found in the endpoints list, return without proceeding further
+	if err != nil {
+		return nil
+	}
+	patchData := fmt.Sprintf("[{\"op\": \"remove\", \"path\": \"/subsets/0/addresses/%s\"}]", nodeIPPosition)
+
+	_, err = pc.k8sclientset.CoreV1().Endpoints(pc.namespace).Patch(context.TODO(),
+		windowsMetricsEndpoints, types.JSONPatchType, []byte(patchData), metav1.PatchOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "could not remove node %v from endpoints object", pc.hostname)
+	}
+	log.Info("prometheus configured", "operation", "remove", "IPAddress", pc.endpointAddress.IP)
+	return nil
+}
+
+// Add will create the Services and Service Monitors to allow the operator export the metrics by using
+// the Prometheus operator
+func Add(ctx context.Context, cfg *rest.Config, namespace string) {
+	// Add to the below struct any other metrics ports you want to expose.
+	servicePorts := []v1.ServicePort{
+		{Port: Port, Name: metricsPortName, Protocol: v1.ProtocolTCP, TargetPort: intstr.IntOrString{Type: intstr.Int, IntVal: Port}},
+	}
+
+	// Create Service object to expose the metrics port(s).
+	service, err := metrics.CreateMetricsService(ctx, cfg, servicePorts)
+	if err != nil {
+		log.Error(err, "could not create metrics Service")
```

Comment:
Why not error out here? It seems wrong to me to continue configuring metrics if such big errors occur. I think returning an error and deciding how handle it at the call site is a better way to approach this. If we dont want metrics failures to stop the operator from coming up thats fine, but we can return these errors and the log them at the call site.

Reply:
This is called in `main.go`. Do you want me to return error and `log` the message there ?

Reply:
I think that would be ideal